### PR TITLE
Make Python local and global scopes available for query formulas

### LINKED ISSUE #1072: Using Python globals() to add objects for use by the query engine doesn't work for sub modules
**Description**
Because Python globals() only provides access to the enclosing module's global symbols not the main module's, it doesn't work from within any sub modules imported.

**@alexpeters1208 ran into this problem with his learn module and had to revert back to using QueryScope in a hacky way. Once we find a permanent solution, we will need to fix his code.**

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #1072

### CODE REVIEW

File: engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java [hunk 1]
```diff
@@ -325,6 +325,10 @@ private ScriptSessionQueryScope(ScriptSession scriptSession) {
         public void putObjectFields(Object object) {
             throw new UnsupportedOperationException();
         }
+
+        public ScriptSession scriptSession() {
+            return this.scriptSession;
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java:330]: No need for `this.` in this case.
```suggestion
            return scriptSession;
```

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java [hunk 2]
```diff
@@ -200,4 +211,12 @@ public static Object convert(PyObject pyObject) {
     public PyDictWrapper globals() {
         return dict;
     }
+
+    @Override
+    public void modifyScope(PyObject pydict) {
+        if (pydict.isNone())
+            moduleDict.set(null);
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java:218]: Could we just simplify this and `currentScope` by just installing `this.dict` as `moduleDict` when `null` is passed?

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java [hunk 3]
```diff
@@ -170,4 +171,6 @@ default Map<String, Object> getEntriesMap() {
     }
 
     public PyDictWrapper globals();
+
+    void modifyScope(PyObject object);
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java:175]: Why not `setScope`? `setScopeDictionary`? `setModuleDict`? Need some JavaDoc.

File: py/server/deephaven/table.py [hunk 4]
```diff
@@ -7,6 +7,8 @@
 
 from __future__ import annotations
 
+import inspect
+import sys
```
Comment (rcaudy) [inline py/server/deephaven/table.py:11]: IntelliJ seems to think this is unnecessary.

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java [hunk 5]
```diff
@@ -200,4 +211,12 @@ public static Object convert(PyObject pyObject) {
     public PyDictWrapper globals() {
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java:211]: I'm a bit nervous about this. `globals()` is what is used to expose named objects to clients... I *think* that's appropriate going forward, but I want to make sure we discuss it. Module dictionaries can be used inside expressions, but putting tables in module dictionaries won't make them visible to clients as-is.

File: py/server/deephaven/table.py [hunk 6]
```diff
@@ -330,6 +352,7 @@ def update(self, formulas: Union[str, Sequence[str]]) -> Table:
         """
         try:
             formulas = to_sequence(formulas)
+            _set_query_scope()
```
Comment (rcaudy) [inline py/server/deephaven/table.py:355]: Don't we need this in a lot more places? All of `view`, `select`, `updateView`, `update`, `lazyUpdate`, `where`, `selectDistinct`, at a minimum.

File: py/server/deephaven/table.py [hunk 7]
```diff
@@ -43,6 +45,26 @@
...
+_JQueryScope = jpy.get_type("io.deephaven.engine.table.lang.QueryScope")
+_JUnsynchronizedScriptSessionQueryScope = jpy.get_type(
+    "io.deephaven.engine.util.AbstractScriptSession$UnsynchronizedScriptSessionQueryScope")
+_JPythonScriptSession = jpy.get_type("io.deephaven.integrations.python.PythonDeephavenSession")
+_j_script_session = jpy.cast(_JQueryScope.getScope(), _JUnsynchronizedScriptSessionQueryScope).scriptSession()
+_j_py_script_session = jpy.cast(_j_script_session, _JPythonScriptSession)
+_deephaven_pkg_prefix = __file__[:__file__.index("deephaven") + len("deephaven")]
+
+
+def _set_query_scope():
```
Comment (rcaudy) [inline py/server/deephaven/table.py:58]: Shouldn't this be a context manager? I think you'd want to install the module dict for the duration of an op, but not leave it installed after.

**Review by rcaudy — COMMENTED** [review]
I haven't tried this out, but I have some commentary on the current state.

File: py/server/deephaven/table.py [hunk 8]
```diff
@@ -352,8 +369,7 @@ def update(self, formulas: Union[str, Sequence[str]]) -> Table:
         """
         try:
             formulas = to_sequence(formulas)
-            _set_query_scope()
-            with auto_locking_ctx(self):
+            with auto_locking_ctx(self), _query_scope_ctx():
```
Comment (rcaudy) [inline py/server/deephaven/table.py:372]: Is there an implied ordering here? I would prefer to hold the lock for the shortest possible window.

File: py/server/deephaven/table.py [hunk 9]
```diff
@@ -52,18 +52,35 @@
...
-_deephaven_pkg_prefix = __file__[:__file__.index("deephaven") + len("deephaven")]
+_deephaven_pkg_prefix = __file__
 
 
-def _set_query_scope():
-    for frame, filename, line_num, func, source_code, source_index in inspect.getouterframes(inspect.currentframe()):
-        if filename and filename.startswith(_deephaven_pkg_prefix):
-            continue
+@contextlib.contextmanager
+def _query_scope_ctx():
```
Comment (rcaudy) [inline py/server/deephaven/table.py:59]: Are we sure this should be internal-only?

File: py/server/tests/test_pandas.py [hunk 10]
```diff
@@ -11,7 +11,7 @@
 from deephaven import dtypes, new_table, DHError
 from deephaven.column import byte_col, char_col, short_col, bool_col, int_col, long_col, float_col, double_col, \
     string_col, datetime_col, pyobj_col, jobj_col
-from deephaven.constants import NULL_LONG, NULL_SHORT, NULL_INT
+from deephaven.constants import NULL_LONG, NULL_SHORT, NULL_INT, NULL_BYTE as NB
```
Comment (rcaudy) [inline py/server/tests/test_pandas.py:14]: Why the `as NB`?

File: py/server/deephaven/table.py [hunk 11]
```diff
@@ -43,6 +45,43 @@
...
+
+@contextlib.contextmanager
+def _query_scope_ctx():
+    """A context manager to set/unset query scope based on the scope of the most immediate application code that invokes
+    Table operations."""
+
+    # locate the innermost Deephaven frame (i.e. any of the table operation methods that use this context manager)
+    outer_frames = inspect.getouterframes(inspect.currentframe())[1:]
+    for i, (frame, filename, *_) in enumerate(outer_frames):
+        if filename and filename == __file__:
```
Comment (rcaudy) [inline py/server/deephaven/table.py:66]: Did you intend to use `_deephaven_pkg_prefix` here?

File: py/server/deephaven/table.py [hunk 12]
```diff
@@ -43,6 +45,43 @@
...
+_j_py_script_session = jpy.cast(_j_script_session, _JPythonScriptSession)
+_deephaven_pkg_prefix = __file__
+
+
+@contextlib.contextmanager
+def _query_scope_ctx():
+    """A context manager to set/unset query scope based on the scope of the most immediate application code that invokes
+    Table operations."""
+
+    # locate the innermost Deephaven frame (i.e. any of the table operation methods that use this context manager)
```
Comment (rcaudy) [inline py/server/deephaven/table.py:63]: Why do we want the innermost Deephaven frame, rather than the outermost Deephaven-only frame? I'm not sure I understand.

**Review by rcaudy — COMMENTED** [review]
I'm happy enough with the current state, modulo my questions/comments. On to @chipkent or @devinrsmith

File: py/server/deephaven/table.py [hunk 13]
```diff
@@ -349,7 +388,7 @@ def lazy_update(self, formulas: Union[str, Sequence[str]]) -> Table:
         """
         try:
             formulas = to_sequence(formulas)
-            with auto_locking_ctx(self):
+            with auto_locking_ctx(self), _query_scope_ctx():
```
Comment (devinrsmith) [inline py/server/deephaven/table.py:391]: How does globals() work wrt defaults? `def my_method(formulas, context=globals())`? Does it capture `globals()` from the callers perspective, or from the currently defined module?

**Review by devinrsmith — COMMENTED** [review]
Snazzy. I like the feel of this - would be nice to follow a similar pattern for groovy!

File: py/server/tests/test_table.py [hunk 14]
```diff
@@ -560,6 +560,47 @@ def test_layout_hints(self):
         self.assertTrue(cm.exception.root_cause)
         self.assertIn("RuntimeError", cm.exception.compact_traceback)
 
+    def test_update_local(self):
```
[the remark this answers is not in the document]
Comment (chipkent) [reply py/server/tests/test_table.py:563]: Should also test what happens when variables change.  e.g.
```python
x = 1
t = emptyTable(1).update("X=x")
assert X==1
x = 2
t = t.update("Y=x")
assert Y==2
```

File: py/server/deephaven/table.py [hunk 15]
```diff
@@ -43,6 +45,43 @@
...
 _JTableOperations = jpy.get_type("io.deephaven.api.TableOperations")
 
+# Dynamic Query Scope
+_JQueryScope = jpy.get_type("io.deephaven.engine.table.lang.QueryScope")
+_JUnsynchronizedScriptSessionQueryScope = jpy.get_type(
+    "io.deephaven.engine.util.AbstractScriptSession$UnsynchronizedScriptSessionQueryScope")
+_JPythonScriptSession = jpy.get_type("io.deephaven.integrations.python.PythonDeephavenSession")
+_j_script_session = jpy.cast(_JQueryScope.getScope(), _JUnsynchronizedScriptSessionQueryScope).scriptSession()
+_j_py_script_session = jpy.cast(_j_script_session, _JPythonScriptSession)
+_deephaven_pkg_prefix = __file__
```
[the remark this answers is not in the document]
Comment (chipkent) [reply py/server/deephaven/table.py:55]: after reading more and seeing the slick way you've made scope work in functions without users having to explicitly create a scope, do you want a private query scope module?

Are there any cases, other than operations on tables, where query scope can be used?  If not, then maybe it makes sense to have the functionality in this file with `__file__` hard coded.

File: py/server/deephaven/table.py (hunk 9 above)
Comment (chipkent) [reply py/server/deephaven/table.py:59]: After reading the tests, it might be totally reasonable for the context to be private, since it looks like @jmao-denver has managed to get function level scope automagically, without users needing to do anything.

File: py/server/deephaven/table.py [hunk 16]
```diff
@@ -43,6 +45,43 @@
...
+
+
+@contextlib.contextmanager
+def _query_scope_ctx():
+    """A context manager to set/unset query scope based on the scope of the most immediate application code that invokes
+    Table operations."""
```
Comment (chipkent) [inline py/server/deephaven/table.py:61]: Users aren't going to be able to understand this.  Consider making the description longer, so that it is understandable.

File: py/server/deephaven/table.py [hunk 17]
```diff
@@ -43,6 +45,43 @@
...
+
+    # locate the innermost Deephaven frame (i.e. any of the table operation methods that use this context manager)
+    outer_frames = inspect.getouterframes(inspect.currentframe())[1:]
+    for i, (frame, filename, *_) in enumerate(outer_frames):
+        if filename and filename == __file__:
+            break
+
+    # we deem any module/function/method which calls a wrapped formula/filter-evaluating Table op as an application
+    # frame. This holds true and makes sense even when such module/function/method is part of the Deephaven package, and
+    # should not be a hurdle for developing such code as function/method parameters are now part of the query scope
```
Comment (chipkent) [inline py/server/deephaven/table.py:71]: not totally clear.  I had to read it multiple times to get an idea of what it is trying to say.

File: py/server/deephaven/table.py [hunk 18]
```diff
@@ -43,6 +45,43 @@
...
+        scope_dict = app_frame.f_globals.copy()
+        scope_dict.update(app_frame.f_locals)
+        try:
+            _j_py_script_session.pushScope(scope_dict)
+            yield
+        finally:
+            _j_py_script_session.popScope()
+    else:
+        # in the __main__ module, use the default main global scope
+        yield
```
Comment (chipkent) [inline py/server/deephaven/table.py:83]: 1) Are function locals in the main global scope handled by the if-block?
2) Are we properly handling values in an `if __name__ == "__main__":` block?

I think we are ok on these, but I'm confirming, since I'm not 100% sure.  Are there any other weird cases to consider?

File: py/server/tests/test_table.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply py/server/tests/test_table.py:563]: Similarly, what happens with `time_table`?  Are we freezing the values?

```python
x = 1
t = time_table(...).update("X=x")
for i in range(1e6):
    x = i
```

File: py/server/deephaven/table.py (hunk 15 above)
Comment (chipkent) [inline py/server/deephaven/table.py:55]: Is `_deephaven_pkg_prefix` needed?  I don't think it is used.

File: py/server/tests/test_table.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply py/server/tests/test_table.py:563]: should also test other scoping cases that are unusual in python:
* lambdas
* list comprehensions
* dict comprehensions
* set comprehensions
* class methods
* static class methods

This link should give a number of weird cases to test against:
https://realpython.com/python-scope-legb-rule/

I'm just paranoid that we check edge cases.

File: py/server/deephaven/table.py [hunk 19]
```diff
@@ -2123,7 +2167,7 @@ def select_distinct(self, formulas: Union[str, Sequence[str]] = None) -> Partiti
         """
         try:
             formulas = to_sequence(formulas)
-            with auto_locking_ctx(self):
+            with auto_locking_ctx(self), _query_scope_ctx():
```
Comment (chipkent) [inline py/server/deephaven/table.py:2170]: Did we remove formulas from joins, aggs, etc., or do they need a scope?

File: py/server/tests/test_table.py [hunk 20]
```diff
@@ -560,6 +560,47 @@ def test_layout_hints(self):
...
+
+        inner_func("param str")
+
+    def test_nested_scopes(self):
+        def inner_func(p) -> str:
+            t = empty_table(1).update("X = p * 10")
+            return t.to_string().split()[2]
+
+        t = empty_table(1).update("X = i").update("TableString = inner_func(X + 10)")
+        self.assertIn("100", t.to_string())
```
Comment (chipkent) [inline py/server/tests/test_table.py:596]: Should test partitioned tables.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/server/deephaven/table.py (hunk 9 above)
Comment (jmao-denver) [reply py/server/deephaven/table.py:59]: it is written for the specific purpose of setting formula/filter evaluation scope, unless we expand these evaluation beyond table ops, it is appropriate to keep it internal for now

File: py/server/tests/test_pandas.py (hunk 10 above)
Comment (jmao-denver) [reply py/server/tests/test_pandas.py:14]: one of the test cases expects the Java constant NULL_BYTE in the formula, this is to work around the naming conflicts between Python and Java. A related ticket is #2306.

File: py/server/deephaven/table.py (hunk 12 above)
Comment (jmao-denver) [reply py/server/deephaven/table.py:63]: The next comment in the code tries to explain it. 
```
   # we deem any module/function/method which calls a wrapped formula/filter-evaluating Table op as an application
   # frame. This holds true and makes sense even when such module/function/method is part of the Deephaven package, and
   # should not be a hurdle for developing such code as function/method parameters are now part of the query scope
```
Basically we wan to find the wrapper function that directly uses this context manager and uses its caller's locals and the module globals as the query scope. I have changed the names from application/app to caller in the code to make it clear. This is less ambiguous than the notion of 'application scope', which in most cases is the same as the immediate caller's.

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java [hunk 21]
```diff
@@ -200,4 +211,22 @@ public static Object convert(PyObject pyObject) {
...
+        scopeStack.push(pydict.asDict());
+    }
+
+    @Override
+    public void popScope() {
+        Deque<PyDictWrapper> scopeStack = threadScopeStack.get();
+        if (scopeStack != null) {
+            scopeStack.pop();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java:231]: Throwing an illegal state exception is sound practice. Since push and pop are on the same thread, checking equality doesn't seem to be necessary.

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java [hunk 22]
```diff
@@ -200,4 +211,22 @@ public static Object convert(PyObject pyObject) {
...
+            threadScopeStack.set(scopeStack);
+        }
+        scopeStack.push(pydict.asDict());
+    }
+
+    @Override
+    public void popScope() {
+        Deque<PyDictWrapper> scopeStack = threadScopeStack.get();
+        if (scopeStack != null) {
+            scopeStack.pop();
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PythonScopeJpyImpl.java:229]: makes sense

File: py/server/deephaven/table.py [hunk 23]
```diff
@@ -43,6 +45,43 @@
...
+
+    # we deem any module/function/method which calls a wrapped formula/filter-evaluating Table op as an application
+    # frame. This holds true and makes sense even when such module/function/method is part of the Deephaven package, and
+    # should not be a hurdle for developing such code as function/method parameters are now part of the query scope
+    if len(outer_frames) > i + 2:
+        app_frame, *_ = outer_frames[i + 1]
+        scope_dict = app_frame.f_globals.copy()
+        scope_dict.update(app_frame.f_locals)
+        try:
+            _j_py_script_session.pushScope(scope_dict)
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/server/deephaven/table.py:77]: I don't think we need to protect the scope_dict by holding on to the GIL. It is a new dict. Although its values could be mutable, it is the nature of a multithreaded application and if some of the values are desired to remain unchanged during a table operation, it falls on the application to guard them with any of the available sync mechanisms.

That said, I believe we might encounter cases in the future where holding on to the GIL for the duration of a wrapped operation is necessary.

File: py/server/deephaven/table.py (hunk 13 above)
Comment (jmao-denver) [reply py/server/deephaven/table.py:391]: always the module where it is defined

File: py/server/deephaven/table.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (devinrsmith) [reply py/server/deephaven/table.py:77]: It's not about protecting scope_dict - it's about ensuring `pushScope` is as fast as possible - as it stands right now, there's potential for somebody else to acquire the GIL and for the return from java pushScope -> python to block.

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonDeephavenSession.java [hunk 24]
```diff
@@ -178,6 +178,14 @@ public <T> T getVariable(String name, T defaultValue) {
                 .orElse(defaultValue);
     }
 
+    public void pushScope(PyObject pydict) {
+        scope.pushScope(pydict);
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply Integrations/src/main/java/io/deephaven/integrations/python/PythonDeephavenSession.java:182]: it doesn't hurt for future

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java [hunk 25]
```diff
@@ -170,4 +171,17 @@ default Map<String, Object> getEntriesMap() {
     }
 
     public PyDictWrapper globals();
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java:173]: will rename it to mainGlobals(). it is still meaningful for the gRPC server as table, plots etc in the main globals are automatically exported to the Web UI.

File: py/server/deephaven/table.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/server/deephaven/table.py:55]: I can't find any case other than in the table module that query scope is applicable. If/when query scope is needed in other modules in the future, we should be able to easily generalize the context manager for broader use.

File: py/server/deephaven/table.py (hunk 19 above)
Comment (jmao-denver) [reply py/server/deephaven/table.py:2170]: I think so. Support formulas in joins, etc. adds minimum convenience but would be outweighed by the added complexity in the UI and in the query engine.

File: py/server/deephaven/table.py (hunk 18 above)
Comment (jmao-denver) [reply py/server/deephaven/table.py:83]: 1. yes
2. yes
3. AFAIK, no

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java [hunk 26]
```diff
@@ -170,7 +170,7 @@ default Map<String, Object> getEntriesMap() {
          */
     }
 
-    public PyDictWrapper globals();
+    public PyDictWrapper mainGlobals();
```
Comment (chipkent) [inline engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java:173]: javadoc.  It looks like you renamed it, but I'm still not certain what "it" is.  `globals` may have been a fine name.  It seems like these might be "query globals".

File: py/server/tests/test_pt_proxy.py [hunk 27]
```diff
@@ -318,6 +318,40 @@ def test_agg_all_by(self):
             for gct in agg_pt_proxy.target.constituent_tables:
                 self.assertGreaterEqual(gct.size, 1)
 
+    def test_query_scope(self):
```
Comment (chipkent) [inline py/server/tests/test_pt_proxy.py:321]: Should the whole set of scope tests from table be here?

File: py/server/deephaven/table.py (hunk 15 above)
Comment (chipkent) [reply py/server/deephaven/table.py:55]: Is this needed?

File: py/server/tests/test_table.py [hunk 28]
```diff
@@ -560,6 +562,109 @@ def test_layout_hints(self):
         self.assertTrue(cm.exception.root_cause)
         self.assertIn("RuntimeError", cm.exception.compact_traceback)
 
+    def verify_table_data(self, t: Table, expected: List[Any]):
+        t_data = to_pandas(t).values[0]
+        for s in expected:
+            self.assertIn(s, t_data)
+
+    def test_update_local(self):
```
Comment (chipkent) [inline py/server/tests/test_table.py:570]: Should also test what happens when variables change. e.g.

```python
x = 1
t = emptyTable(1).update("X=x")
assert X==1
x = 2
t = t.update("Y=x")
assert Y==2
```

Comment (chipkent) [reply py/server/tests/test_table.py:570]: Similarly, what happens with time_table? Are we freezing the values?
```python
x = 1
t = time_table(...).update("X=x")
for i in range(1e6):
    x = i
```

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/server/tests/test_pt_proxy.py (hunk 27 above)
Comment (jmao-denver) [reply py/server/tests/test_pt_proxy.py:321]: I don't think it is necessary. The new tests for table cover most if not all the use cases to verify that we build the correct scope for formula evaluation

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java (hunk 26 above)
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java:173]: it is the Python's main module namespace as opposed to any imported module's.

File: py/server/deephaven/table.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/server/deephaven/table.py:75]: I don't believe the performance argument is as strong as the consistency one, but filed #2741

File: py/server/tests/test_pt_proxy.py (hunk 27 above)
Comment (chipkent) [reply py/server/tests/test_pt_proxy.py:321]: I suggested this because @rcaudy implied that partitioned tables were somehow different than other cases.

**Review by chipkent — APPROVED** [review]
See questions in chat for @rcaudy
